### ISSUE TITLE
Add some unit tests?

### ISSUE THREAD

**Issue by gafferongames** [KillaMaaki/Netcode.IO.NET#4]
One of the things I noticed when I was developing the C implementation was that unit tests helped me avoid breaking things.

As the C# implementation starts to mature, can I suggest adding a bunch of unit tests and perhaps a soak test (repeated client/server connects and disconnects), would be a really good way to flush out issues, and also to catch any errors if somebody checks in something.

eg. you can run the unit tests on each checkin via travis.

Has worked great so far for the C implementation.

cheers

- Glenn

**Comment by gafferongames** [comment]
Another thing that might be useful, is when you have a C# client, test that it connects and works properly against the C implementation of a server and vice versa. This is a good way to ensure your C# implementation matches the spec. Similarly, this would be a good thing to run on each commit via travis.

**Comment by GlaireDaggers** [comment]
Yeah, I'd thought about going through and porting over the unit tests from the C implementation. I may have to modify some things to add test methods, but should help ensure it's as functional as I think it is, and that I don't keep accidentally breaking things going forward ;)

As far as testing with the C implementation, I have been testing with the C# wrapper around the native implementation, having my C# client connect to a wrapped server. It's so far been the easiest way to test, actually. Although, I have yet to test the other way around, which I definitely should.

**Comment by gafferongames** [comment]
I think this is a great idea

**Comment by GlaireDaggers** [comment]
Actually, I guess the thing I said about not testing against a reference client isn't strictly true - I DID test a WebGL Unity build against a .NET server, and the WebGL Unity build of course used the browser extension, which itself is implemented as a wrapper around the reference implementation. So that did work.

Anyway, I've got a bunch of unit tests ported over from netcode.io, and that did help me resolve a few bugs and make some important API changes. I've also got a test method which just runs all of the connection tests in a while loop with some duration. I've so far only tested it set to 10 minutes (around 1500 iterations of all connection test cases on my machine), though tonight perhaps when I go to bed I'll set it to something _significantly_ longer and check the results in the morning :)
Also probably tomorrow I'll put together a more involved soak test based on the soak.c test in the reference implementation.
